### PR TITLE
GUIFormspecMenu: Add animated_image[] element

### PR DESCRIPTION
**Pull request by kilbith**
This PR offered by [Kidscode](https://kidscode.com) adds a new formspec element: `animated_image`

**Demo video:** https://www.youtube.com/watch?v=mDjExI7xBZA

### CODE REVIEW

**Comment by paramat** [comment]
What clang format suggests is useless, your formatting is better, so ignore that build check =)
Please add the file to the clang format whitelist to silence it https://github.com/minetest/minetest/blob/master/util/travis/clang-format-whitelist.txt

File: src/gui/guiFormSpecMenu.cpp [hunk 1]
```diff
@@ -3023,8 +3065,42 @@ void GUIFormSpecMenu::drawMenu()
...
+		Draw animated images
+	*/
+	m_texture_pool.step();
+
+	for (const GUIFormSpecMenu::ImageDrawSpec &spec : m_animated_images) {
+		s32 idx = spec.texture_idx;
+		video::ITexture *texture = m_texture_pool.getTexture(
+				spec.name, m_tsrc, idx);
+
+		if (texture != 0) {
```
Comment (nerzhul) [inline src/gui/guiFormSpecMenu.cpp]: if (texture != NULL)

File: src/gui/guiFormSpecMenu.cpp [hunk 2]
```diff
@@ -3023,8 +3065,42 @@ void GUIFormSpecMenu::drawMenu()
...
+			const core::dimension2d<u32> &img_origsize = texture->getOriginalSize();
+			// Image size on screen
+			core::rect<s32> imgrect;
+
+			if (spec.scale)
+				imgrect = core::rect<s32>(0,0, spec.geom.X, spec.geom.Y);
```
Comment (nerzhul) [inline src/gui/guiFormSpecMenu.cpp]: space issue 0, 0,

File: src/client/texture_pool.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#pragma once
+
+#include "texture.h"
+#include <string>
+#include <vector>
+#include <iostream>
+
+struct TexturePool
+{
+	TexturePool() {}
```
Comment (nerzhul) [inline src/client/texture_pool.h]: TexturePool = default;

File: src/client/texture_pool.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+	TexturePool(const TexturePool &texture_pool) :
+		textures(texture_pool.textures),
+		animations(texture_pool.animations),
+		global_time(texture_pool.global_time)
+	{
+	}
+
+	~TexturePool() {}
```
Comment (nerzhul) [inline src/client/texture_pool.h]: ~TexturePool = default;

File: src/client/texture_pool.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	animations.push_back(texture_idx - 1);
+
+	return texture_idx;
+}
+
+s32 TexturePool::addTexture(const std::string &name)
+{
+	s32 texture_idx = textures.size() + 1;
+	textures.resize(texture_idx);
+	textures[texture_idx - 1].set(name);
```
Comment (nerzhul) [inline src/client/texture_pool.cpp]: instead of resize + set => push_back

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: src/gui/guiFormSpecMenu.cpp (hunk 1 above)
Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]: or just `if (texture)`

File: src/client/texture_pool.cpp (hunk 5 above)
Comment (kilbith) [reply src/client/texture_pool.cpp:37]: This is not `std::set` but `Texture::set`

**Comment by SmallJoker** [comment]
Aside the rebase, I think it would make sense to draw the elements in the order they are parsed (as a follow-up to #8740).

**Comment by kilbith** [comment]
The rebase won't be done by myself.

**Comment by Df458** [comment]
I've rebased locally. I haven't converted `animated_image` over to be a proper element yet (so `ImageDrawSpec` is temporarily back) but I'm probably going to sit down and do that this weekend.

Once I've done that, what does process of getting my changes into this PR look like?

**Review by rubenwardy — COMMENTED** [review]
Looks good, but needs rebasing to move to a GUIElement

Edit: just read the comments.

@Df458: I suggest making a new PR

**Comment by Df458** [comment]
See #9258

**Pull request closed without merging**